### ISSUE TITLE
Bug: error TS2742 in eslint.config.mjs with `"composite": true` and pnpm

### ISSUE THREAD

**Issue by andersk** [typescript-eslint/typescript-eslint#10893]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

In a project configured with `"composite": true` and installed with pnpm, TypeScript flags error TS2742 in this minimal `eslint.config.mjs`.

```js
// @ts-check
import tseslint from "typescript-eslint";
export default tseslint.config(tseslint.configs.recommended);
```

Full `tsconfig.json`:

```json
{
  "compilerOptions": {
    "allowJs": true,
    "composite": true,
    "module": "nodenext",
    "noEmit": true,
    "strict": true
  }
}
```

The error is not shown without `"composite": true`, nor is it shown when using npm instead of pnpm.

Related:

- #10508
- #10513 (merged in v8.18.2)

### Reproduction Repository Link

https://gist.github.com/andersk/44d8dd0d3c2508f758bf51e5cbf12c2f

### Repro Steps

```console
$ git clone https://gist.github.com/andersk/44d8dd0d3c2508f758bf51e5cbf12c2f test
$ cd test
$ pnpm i
Packages: +116
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 116, reused 116, downloaded 0, added 116, done

dependencies:
+ eslint 9.21.0
+ typescript 5.7.3
+ typescript-eslint 8.25.0

Done in 1.9s using pnpm v10.5.1

$ pnpm exec tsc
eslint.config.mjs:3:1 - error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/@typescript-eslint+utils@8.25.0_eslint@9.21.0_typescript@5.7.3/node_modules/@typescript-eslint/utils/ts-eslint'. This is likely not portable. A type annotation is necessary.

3 export default tseslint.config(tseslint.configs.recommended);
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in eslint.config.mjs:3
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.25.0` |
| `@typescript-eslint/parser`            | `8.25.0` |
| `@typescript-eslint/scope-manager`     | `8.25.0` |
| `@typescript-eslint/typescript-estree` | `8.25.0` |
| `@typescript-eslint/type-utils`        | `8.25.0` |
| `@typescript-eslint/utils`             | `8.25.0` |
| `TypeScript`                           | `5.7.3` |
| `ESLint`                               | `9.21.0` |
| `node`                                 | `22.14.0` |

**Comment by bradzacher** [comment]
This is very similar to an issue that eslint upstream also had. Pnpm's linking installs plays a bit funky with typescript so it needs the package to declare some additional type exports for it to work.

**Comment by yuyiz67** [comment]
Hi, I'd like to contribute and work on this issue. Based on my initial research, this might be related to TypeScript's handling of module resolution with pnpm. I'll investigate possible fixes, such as explicitly exporting the required types or adjusting the tsconfig settings. Let me know if you have any suggestions. Thanks!

**Comment by fdendorfer** [comment]
I had the same error, but using a slightly different setup without tseslint.config(). I solved it by adding "eslint.config.mjs" to the exclude array in my tsconfig.json.

**Comment by ntnyq** [comment]
Upstream related issue: 

https://github.com/eslint/eslint/issues/19421

ESLint recommended workaround:
https://eslint.org/docs/latest/use/getting-started#manual-set-up

**Comment by fdendorfer** [comment]
Thanks for the hint @ntnyq 

I can confirm this also works for me

**Comment by michael-farah** [comment]
Hi, I'd like to take this on as my first contribution! Could you please assign it to me?

**Comment by ntdiary** [comment]
Hi, I've looked into this `good first issue` this week, and hope the following information helps :)
Here is the root cause analysis:
1. `composite: true` [implies](https://www.typescriptlang.org/tsconfig/#composite) `declaration: true` by default, which makes typescript assume you'll generate a `.d.mts` for `eslint.config.mjs`.
2. the default [`node-linker=isolated`](https://pnpm.io/9.x/npmrc#node-linker) means dependencies are **symlinked** from the `.pnpm` directory.
3. since `@typescript-eslint/utils` isn't in the project's **runtime dependencies** (i.e.,`package.json`), typescript can only recognize the real path that has a nested `node_modules`, rather than convert it to its simpler **symlink** path, which finally leads to the TS2742 error[^pic].
    ```shell
    @typescript-eslint/utils -> ../.pnpm/@typescript-eslint+utils@xxx/node_modules/@typescript-eslint/utils
    ```

Based on those, here are a few ways to address this issue—(just select the one suits your project):
1. **exclude `eslint.config.mjs` in your `tsconfig.json`**
    This lets typescript know you don’t need the `eslint.config.d.mts` file, you can use either `include` or `exclude` (might need to be aware that the latter has [default values](https://www.typescriptlang.org/tsconfig/#exclude)).
    ```jsonc
    "include": [
      // only include ts files
      "*.ts",
    ]
    ```
    ```jsonc
    "exclude": [
      // other items,
      "eslint.config.mjs",
    ],
    ```
2. **set `node-linker=hoisted` in your .npmrc** (as [previous comment](https://github.com/typescript-eslint/typescript-eslint/issues/10893#issuecomment-2722391138) suggested)
  you will get a **flat** `node_modules` without symlinks are created, just like npm or yarn.

3. **run `pnpm install @typescript-eslint/utils`** to make this dependency explicit.
  this allows typescript to recognize its symlink mentioned above (as typescript automatically reads the `dependencies` in `package.json`)

4. **export `FlatConfig` namespace from `typescript-eslint`**
  this package is already in the `dependencies`, so typescript can correctly recognize its symlink. BTW, I saw that exporting `FlatConfig` was approved in issue [#10848](https://github.com/typescript-eslint/typescript-eslint/issues/10848), and that  PR hasn’t been updated yet, so not sure about what next steps will be, but I’d be happy to send a PR if needed.


[^pic]: [source code image](https://github.com/user-attachments/assets/81914580-5677-41de-a0c4-0048c5f0b0ff)

**Comment by ntdiary** [comment]
#10849 also fixed this issue. :)
cc @JoshuaKGoldberg

**Comment by ritishDas** [comment]
Hello,
I noticed that this issue seems to be fixed in the current main branch — I tested it locally, and the behavior now works as expected.
Could you please confirm if this can be closed?
Thanks!

**Comment by kirkwaiblinger** [comment]
Confirmed this does not reproduce on typescript-eslint@8.46.2
